applications/ShapeOptimizationApplication/custom_utilities/response_functions/water_drain_response_function_utility.cpp

```diff
@@ -0,0 +1,454 @@
+					break;
+				}
+			}
+		}
+
+		if (low_point) {
+			Volume new_volume;
+			new_volume.Id = volume_id;
+			rNode.FastGetSolutionStepValue(WATER_VOLUMES) = new_volume.Id;
+			new_volume.mLowPointId = rNode.Id();
+			new_volume.mLowestPoint = Vector(rNode.Coordinates());
+			mListOfVolumes.push_back(new_volume);
+			++volume_id;
+		}
+	}
+
+	KRATOS_INFO("ShapeOpt::WaterDrain") << "Finished SearchLowPoints in " << timer.ElapsedSeconds() << " s." << std::endl;
+}
+
+void WaterDrainResponseFunctionUtility::FindSteepestDescentFromEachNode() {
+
+	KRATOS_TRY;
+
+	BuiltinTimer timer;
+
+	ModelPart& r_edge_model_part = mrModelPart.GetModel().GetModelPart(mFreeEdgeSubModelPartName);
+
+	KRATOS_INFO("ShapeOpt::WaterDrain") << "Starting FindSteepestDescentFromEachNode..." << std::endl;
+	block_for_each(mrModelPart.Nodes(), [&](Node& rNode) {
+		IndexType current_node_id = rNode.Id();
+
+		if (!r_edge_model_part.HasNode(current_node_id)) {
+			bool is_low_point = false;
+			for (std::vector<Kratos::Volume>::size_type i = 0; i<mListOfVolumes.size(); i++) {
+				if (current_node_id == mListOfVolumes[i].mLowPointId) {
+					is_low_point = true;
+					break;
+				}
+			}
+			if (!is_low_point) {
+				int iter = 0;
+				bool low_point_found = false;
+				// search in steepest descent direction until low point or free edge is found
+				std::set<int> nodes_visited;
+				while (!low_point_found && iter < mMaxIterations) {
+
+					if (!nodes_visited.insert(current_node_id).second) break;
+
+					NodeTypePointer p_current_node = mrModelPart.pGetNode(current_node_id);
+					auto& r_neighbours = p_current_node->GetValue(NEIGHBOUR_NODES);
+
+					IndexType next_node_id;
+					double max_inclination = 0;
+					for (auto& r_neighbour : r_neighbours) {
+						const auto& distance_vector = r_neighbour.Coordinates() - p_current_node->Coordinates();
+						const double inclination = MathUtils<double>::Dot3(mGravityDirection, distance_vector);
+						if (inclination > max_inclination) {
+							max_inclination = inclination;
+							next_node_id = r_neighbour.Id();
+						}
+					}
+					if (max_inclination < 1e-8) break;
+
+					if (r_edge_model_part.HasNode(next_node_id)) break;
+
+					for (std::vector<Kratos::Volume>::size_type i = 0; i<mListOfVolumes.size(); i++) {
+						if (next_node_id == mListOfVolumes[i].mLowPointId) {
+							rNode.FastGetSolutionStepValue(WATER_VOLUMES) = mListOfVolumes[i].Id;
+							low_point_found = true;
+							break;
+						}
+					}
+					current_node_id = next_node_id;
+					++iter;
+				}
+			}
+		}
+	});
+
+	KRATOS_INFO("ShapeOpt::WaterDrain") << "Finished FindSteepestDescentFromEachNode in " << timer.ElapsedSeconds() << " s." << std::endl;
+
+	KRATOS_CATCH("");
+}
+
+void WaterDrainResponseFunctionUtility::LevelVolumes() {
+
+	KRATOS_TRY;
+
+	BuiltinTimer timer;
+	KRATOS_INFO("ShapeOpt::WaterDrain") << "Starting LeveVolumes..." << std::endl;
+
+	for (std::vector<Kratos::Volume>::size_type i = 0; i<mListOfVolumes.size(); i++) {
```

Comment:
Id it possible to parallelize such loops?

Reply:
I've parallelized the nested loop over the nodes inside this loop. The number of volumes/ponds is usually much smaller than the amount of nodes.

fyi: the response takes in total around 0.5s for a model with 120000 nodes. That is much faster than the matrix free vm mapper needs to do the mapping.